third_party/address_sorting/address_sorting.c

```diff
@@ -0,0 +1,405 @@
+
+static address_sorting_socket_factory* g_current_socket_factory = NULL;
+
+static int default_socket_factory_socket(address_sorting_socket_factory* self,
+                                         int domain, int type, int protocol) {
+  return socket(domain, type, protocol);
+}
+
+static int default_socket_factory_connect(address_sorting_socket_factory* self,
+                                          int sockfd,
+                                          const struct sockaddr* addr,
+                                          socklen_t addrlen) {
+  return connect(sockfd, addr, addrlen);
+}
+
+static int default_socket_factory_getsockname(
+    address_sorting_socket_factory* self, int sockfd, struct sockaddr* addr,
+    socklen_t* addrlen) {
+  return getsockname(sockfd, addr, addrlen);
+}
+
+static int default_socket_factory_close(address_sorting_socket_factory* self,
+                                        int sockfd) {
+  return close(sockfd);
+}
+
+static void default_socket_factory_destroy(
+    address_sorting_socket_factory* self) {
+  return;
+}
+
+const address_sorting_socket_factory_vtable default_socket_factory_vtable = {
+    default_socket_factory_socket,      default_socket_factory_connect,
+    default_socket_factory_getsockname, default_socket_factory_close,
+    default_socket_factory_destroy,
+};
+
+static address_sorting_socket_factory* create_default_socket_factory() {
+  address_sorting_socket_factory* factory =
+      gpr_zalloc(sizeof(address_sorting_socket_factory));
+  factory->vtable = &default_socket_factory_vtable;
+  return factory;
+}
+
+static int address_sorting_socket(int domain, int type, int protocol) {
+  return g_current_socket_factory->vtable->socket(g_current_socket_factory,
+                                                  domain, type, protocol);
+}
+
+static int address_sorting_connect(int sockfd, const struct sockaddr* addr,
+                                   socklen_t addrlen) {
+  return g_current_socket_factory->vtable->connect(g_current_socket_factory,
+                                                   sockfd, addr, addrlen);
+}
+
+static int address_sorting_getsockname(int sockfd, struct sockaddr* addr,
+                                       socklen_t* addrlen) {
+  return g_current_socket_factory->vtable->getsockname(g_current_socket_factory,
+                                                       sockfd, addr, addrlen);
+}
+
+static int address_sorting_close(int sockfd) {
+  return g_current_socket_factory->vtable->close(g_current_socket_factory,
+                                                 sockfd);
+}
+
+static int ipv6_prefix_match_length(const struct sockaddr_in6* sa,
+                                    const struct sockaddr_in6* sb) {
+  unsigned char* a = (unsigned char*)&sa->sin6_addr;
+  unsigned char* b = (unsigned char*)&sb->sin6_addr;
+  int cur_bit = 0;
+  while (cur_bit < 128) {
+    int a_val = a[cur_bit / CHAR_BIT] & (1 << (cur_bit % CHAR_BIT));
```

Comment:
This algorithm looks wrong.  Assume there are 15 bytes, each containing 8 bits:

    [0]:76543210, [1]:76543210, ..., [15]:76543210

Your code checks `[0]:0, [0]:1, ...` instead of `[0]:7, [0]:6, ...`.
I recommend making the test fail before fixing the code.

Reply:
Fixed and added `TestPrefersLongestPrefixStressInnerBytePrefix` to catch this. Good catch, thanks!